import grpc
from datetime import datetime
from infrastructure.grpc.generated import eleicao_pb2, eleicao_pb2_grpc

def create_eleicao(stub):
    print("‚û°Ô∏è Criando uma nova elei√ß√£o...")
    request = eleicao_pb2.CreateEleicaoRequest(
        titulo="Mais uma Eleicao Teste 10",
        descricao="Criada pelo client gRPC",
        data_start=datetime(2025, 7, 1).isoformat(),
        data_final=datetime(2025, 7, 10).isoformat()
    )

    try:
        response = stub.CreateEleicao(request)
        print("‚úÖ Elei√ß√£o criada com sucesso:")
        print(f"ID: {response.id}")
        print(f"T√≠tulo: {response.titulo}")
        print(f"Status: {response.status}")
        return response.id  # Retorna o ID para buscar depois
    except grpc.RpcError as e:
        print("‚ùå Erro ao criar elei√ß√£o:")
        print(f"Status: {e.code().name}")
        print(f"Detalhes: {e.details()}")
        return None

def get_eleicao(stub, eleicao_id):
    print(f"\nüîç Buscando elei√ß√£o com ID: {eleicao_id}")
    request = eleicao_pb2.EleicaoRequest(id=eleicao_id)

    try:
        response = stub.GetEleicao(request)
        print("‚úÖ Elei√ß√£o encontrada:")
        print(f"ID: {response.id}")
        print(f"T√≠tulo: {response.titulo}")
        print(f"Descri√ß√£o: {response.descricao}")
        print(f"Data In√≠cio: {response.data_start}")
        print(f"Data Final: {response.data_final}")
        print(f"Status: {response.status}")
    except grpc.RpcError as e:
        print("‚ùå Erro ao buscar elei√ß√£o:")
        print(f"Status: {e.code().name}")
        print(f"Detalhes: {e.details()}")

def edit_eleicao(stub, eleicao_id):
    print(f"\n‚úèÔ∏è Editando elei√ß√£o com ID: {eleicao_id}")

    request = eleicao_pb2.EditEleicaoRequest(
        id=eleicao_id,
        titulo="Elei√ß√£o Editada via Client",
        descricao="Agora com nova descri√ß√£o",
        data_start=datetime(2025, 7, 5).isoformat(),
        data_final=datetime(2025, 7, 15).isoformat(),
        status="EM_ANDAMENTO"
    )

    try:
        response = stub.EditEleicao(request)
        print("‚úÖ Elei√ß√£o editada com sucesso:")
        print(f"ID: {response.id}")
        print(f"T√≠tulo: {response.titulo}")
        print(f"Status: {response.status}")
    except grpc.RpcError as e:
        print("‚ùå Erro ao editar elei√ß√£o:")
        print(f"Status: {e.code().name}")
        print(f"Detalhes: {e.details()}")
        
def delete_eleicao(stub, eleicao_id):
    print(f"\nüóëÔ∏è Deletando elei√ß√£o com ID: {eleicao_id}")

    request = eleicao_pb2.DeleteEleicaoRequest(id=eleicao_id)
    try:
        response = stub.DeleteEleicao(request)
        print("‚úÖ", response.message)
    except grpc.RpcError as e:
        print("‚ùå Erro ao deletar elei√ß√£o:")
        print(f"Status: {e.code().name}")
        print(f"Detalhes: {e.details()}")
        
def list_eleicoes(stub):
    print("\nüìã Listando todas as elei√ß√µes...")

    request = eleicao_pb2.ListEleicoesRequest()
    try:
        response = stub.ListEleicoes(request)
        if not response.eleicoes:
            print("‚ö†Ô∏è Nenhuma elei√ß√£o cadastrada.")
        for e in response.eleicoes:
            print(f"- {e.id} | {e.titulo} | {e.status}")
    except grpc.RpcError as e:
        print("‚ùå Erro ao listar elei√ß√µes:")
        print(f"Status: {e.code().name}")
        print(f"Detalhes: {e.details()}")

def run():
    with grpc.insecure_channel("localhost:50051") as channel:
        stub = eleicao_pb2_grpc.EleicaoServiceStub(channel)
        
        # 1. Cria a elei√ß√£o
        eleicao_id = create_eleicao(stub)
        
        #2. Busca a elei√ß√£o
        if eleicao_id:
            get_eleicao(stub, eleicao_id)
            #delete_eleicao(stub, eleicao_id)
            #get_eleicao(stub, eleicao_id)
            
        # list_eleicoes(stub)

if __name__ == "__main__":
    run()
